File: scripts/qa_m1.py
"""M1 QA (lean): verify page renders, canvas mounts, portfolio content present,
and capture boot + a mid-scroll screenshot. Avoids forcing swiftshader (uses
real GPU if available, else the app's own WebGL fallback path)."""
import json
from playwright.sync_api import sync_playwright, Error as PWError

URL = "http://127.0.0.1:4321/"

def main():
    errors, warnings = [], []
    with sync_playwright() as p:
        browser = p.chromium.launch(args=["--disable-dev-shm-usage"])
        page = browser.new_page(viewport={"width": 1440, "height": 900})
        page.set_default_timeout(15000)
        page.on("console", lambda m: (errors.append(m.text) if m.type == "error"
                                      else warnings.append(m.text) if m.type == "warning" else None))
        page.on("pageerror", lambda e: errors.append("PAGEERROR: " + str(e)))

        page.goto(URL, wait_until="domcontentloaded", timeout=15000)
        page.wait_for_timeout(3500)

        has_canvas = page.evaluate("""() => {
            const c = document.querySelector('canvas');
            return c ? {w:c.width,h:c.height} : null;
        }""")
        boot_identity = page.evaluate(
            "document.body.innerText.includes('NAVEEN BISHNOI') ? 'identity-visible' : 'no-identity'")

        # Scroll into the city scene and screenshot
        page.evaluate("""() => {
            const max = document.body.scrollHeight - window.innerHeight;
            window.scrollTo(0, max * 0.8);
        }""")
        page.wait_for_timeout(1500)
        try:
            page.screenshot(path="/tmp/m1_city.png", timeout=8000, animations="disabled")
        except Exception as e:
            print("SHOT_CITY_SKIP:", str(e)[:80])

        # Scroll to bottom -> editorial portfolio should be revealed
        page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
        page.wait_for_timeout(2000)
        portfolio_visible = page.evaluate("""() => {
            const t = document.body.innerText;
            const has = ['Architecture Built for Determinism','Featured Engineering',
                         'Systems Lab','Direct Communication'];
            return has.filter(h => t.includes(h));
        }""")
        try:
            page.screenshot(path="/tmp/m1_portfolio.png", timeout=8000, animations="disabled")
        except Exception as e:
            print("SHOT_PORTFOLIO_SKIP:", str(e)[:80])

        # Assert M2 UI controls exist
        sound_toggle = page.evaluate("""() => {
            const b = Array.from(document.querySelectorAll('button')).find(x => /SOUND (ON|OFF)/.test(x.textContent||''));
            return b ? (b.getAttribute('aria-pressed')==='true'?'on':'off') : 'missing';
        }""")
        q_badge = page.evaluate("""() => {
            const el = Array.from(document.querySelectorAll('span')).find(x => /^Q:(high|medium|low)$/.test((x.textContent||'').trim()));
            return el ? (el.textContent||'').trim() : 'missing';
        }""")
        city_dests = page.evaluate("""() => {
            window.scrollTo(0, (document.body.scrollHeight-window.innerHeight)*0.8);
            return 'scroll-set';
        }""")
        page.wait_for_timeout(2500)
        dest_btns = page.evaluate("""() => Array.from(document.querySelectorAll('button'))
            .map(b=>(b.textContent||'').trim())
            .filter(t=>/PROJECTS|AI LAB|RESUME|CONTACT/.test(t))""")

        print("CANVAS:", json.dumps(has_canvas))
        print("BOOT_IDENTITY:", boot_identity)
        print("SOUND_TOGGLE:", sound_toggle)
        print("Q_BADGE:", q_badge)
        print("CITY_DEST_BTNS:", dest_btns)
        print("PORTFOLIO_SECTIONS_FOUND:", json.dumps(portfolio_visible))
        print("ERROR_COUNT:", len(errors))
        print("ERRORS:", json.dumps(errors[:10]))
        print("WARN_COUNT:", len(warnings))
        browser.close()

if __name__ == "__main__":
    main()

File: scripts/qa_m1.mjs
// M1 QA: headless browser verification of the cinematic camera journey.
// Loads the portfolio, scrolls through the full 700vh track, and samples:
//  - console errors / page errors
//  - presence of the WebGL <canvas>
//  - camera world position (read from the R3F canvas via window.__CINE for proof of 3D travel)
//  - scene HUD identity text at boot
//  - screenshots at several progress points for visual inspection

import { chromium } from 'playwright';

const URL = 'http://127.0.0.1:4321/';
const OUT = '/tmp/m1';

const errors = [];
const warnings = [];

const browser = await chromium.launch({ args: ['--use-gl=swiftshader', '--enable-webgl', '--ignore-gpu-blocklist'] });
const page = await browser.newPage({ viewport: { width: 1440, height: 900 } });

page.on('console', (msg) => {
  if (msg.type() === 'error') errors.push(msg.text());
  if (msg.type() === 'warning') warnings.push(msg.text());
});
page.on('pageerror', (err) => errors.push('PAGEERROR: ' + err.message));

await page.goto(URL, { waitUntil: 'load' });
await page.waitForTimeout(3000);

const hasCanvas = await page.evaluate(() => {
  const c = document.querySelector('canvas');
  return c ? { w: c.width, h: c.height } : null;
});

const bootIdentity = await page.evaluate(() =>
  document.body.innerText.includes('NAVEEN BISHNOI') ? 'identity-visible' : 'no-identity'
);

// Expose camera world position if the R3F store is reachable.
// CinematicExperience uses a zustand store; we instead read the live camera
// by patching into the canvas's __r3f object if present.
async function sampleProgress(pct) {
  const track = await page.evaluate(() => document.body.scrollHeight);
  const y = Math.round((track - window.innerHeight) * pct);
  await page.evaluate((yy) => window.scrollTo(0, yy), y);
  await page.waitForTimeout(900); // let scrub + lerp settle

  const cam = await page.evaluate(() => {
    const c = document.querySelector('canvas');
    // @ts-ignore
    const r3f = c && c.__r3f;
    if (r3f && r3f.root && r3f.root.camera) {
      const cam = r3f.root.camera;
      return { x: +cam.position.x.toFixed(2), y: +cam.position.y.toFixed(2), z: +cam.position.z.toFixed(2) };
    }
    return null;
  });

  const hudScene = await page.evaluate(() => {
    const m = document.body.innerText.match(/SCENE 0\d \/\/ [A-Z ]+/);
    return m ? m[0] : 'none';
  });

  return { pct, cam, hudScene };
}

const samples = [];
for (const pct of [0, 0.15, 0.3, 0.45, 0.6, 0.75, 0.9, 1.0]) {
  const s = await sampleProgress(pct);
  samples.push(s);
  await page.screenshot({ path: `${OUT}_${Math.round(pct * 100)}.png` });
}

// Reverse scroll check
await page.evaluate(() => window.scrollTo(0, 0));
await page.waitForTimeout(800);
const reverseTop = await page.evaluate(() => {
  const c = document.querySelector('canvas');
  // @ts-ignore
  const cam = c && c.__r3f && c.__r3f.root && c.__r3f.root.camera;
  return cam ? { x: +cam.position.x.toFixed(2), y: +cam.position.y.toFixed(2), z: +cam.position.z.toFixed(2) } : null;
});

console.log('CANVAS:', JSON.stringify(hasCanvas));
console.log('BOOT_IDENTITY:', bootIdentity);
console.log('SAMPLES:');
for (const s of samples) console.log(' ', JSON.stringify(s));
console.log('REVERSE_TOP_CAM:', JSON.stringify(reverseTop));
console.log('ERROR_COUNT:', errors.length);
console.log('ERRORS:', JSON.stringify(errors.slice(0, 10)));
console.log('WARN_COUNT:', warnings.length);

await browser.close();
